### PR TITLE
Abort on potential infinite recursion in type_must_outlive (#25954)

### PR DESCRIPTION
**Pull request by robertg**
Display E0399 and stop recursing in `type_must_outlive` if a previous duplicate function application was encountered, to prevent a stack overflow.

This is my initial approach: Declare `(ty, region, origin.span())` to define a distinct `type_must_outlive` function application, and use a `HashSet` to ensure no duplicate function applications occur.

Closes: #25954

### CODE REVIEW

**Comment by arielb1** [comment]
Test please. I'm quite sure you need a recursion limit, not a loop check (the example creates a non-regular type via upvars).

**Comment by robertg** [comment]
@arielb1 I added a compile-fail-test. Are there some examples in the codebase utilizing recursion limits (does it use a function depth constant?). I think figuring out how to detect a loop in the 'region type graph' should be our goal, with using a recursion limit as a last resort.

File: src/librustc_typeck/check/regionck.rs [hunk 1]
```diff
@@ -95,6 +95,7 @@ use middle::ty::{self, ClosureTyper, ReScope, Ty, MethodCall};
 use middle::infer::{self, GenericKind};
 use middle::pat_util;
 
+use std::collections::HashSet;
```
Comment (arielb1) [inline src/librustc_typeck/check/regionck.rs]: Use `FnvHashSet` (that's `util::nodemap::FnvHashSet`).

File: src/test/compile-fail/regionck-unbounded-recursion-type-must-outlive.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// except according to those terms.
+
+// Prevent unbounded recursion during a type_must_outlive pass
+// in regionck.
+
+struct B<F: Fn()> (F);
+
+fn main() {
+    let mut p = B(std::mem::zeroed());
+    p.0 = ||{p.0;} ;
```
Comment (arielb1) [inline src/test/compile-fail/regionck-unbounded-recursion-type-must-outlive.rs]: You need `//~ ERROR error-message` there.

**Comment by arielb1** [comment]
As an approach, I would prefer to check for closure representability (see `ty::check_representable`) rather than this explicit check.

`type_must_outlive` needs to be coinductive, through (is this right, @nikomatsakis?) - this should regionck
```Rust
struct B<'a, F: Fn()+'a>(&'a F);

fn main() {
 let mut p = B(std::mem::zeroed());
 p.0 = &||{p.0;};
}
```
as it is a non-borrowcking sugaring of
```Rust
#![feature(unboxed_closures,core)]

use std::cell::Cell;

struct B<'a, F: FnOnce()+'a>(Cell<Option<&'a F>>);

struct Closure<'a> {
    arg: B<'a, Closure<'a>>
}

impl<'a> FnOnce<()> for Closure<'a> {
    type Output = ();
    extern "rust-call" fn call_once(self, args: ()) {}
}

fn main() {
 let (mut p, mut cls);
 p = B(Cell::new(None));
 cls = Closure { arg: p };
 cls.arg.0.set(Some(&cls));
}
```

**Comment by nikomatsakis** [comment]
Sorry, last week was the work week, and this week I'm on vacation. I'll try to read over the code/comments this week though. I will say that one of the RFCs I am hoping to propose once I get back makes this point about recursion moot.

**Comment by nikomatsakis** [comment]
So, I'm in the process of preparing an RFC and PR that does a big overhaul to `type_must_outlive`, with the aim of fixing some soundness problems. As part of that, it makes infinite recursion impossible. Therefore, I'm inclined to close this PR in favor of that solution. Seem OK?

**Comment by robertg** [comment]
Yes.

**Comment by nikomatsakis** [comment]
Hmm, actually, I was mistaken. The problem this PR aims to correct is not fixed by the patches I've been working on, or at least not the ones I was thining of. However, it IS fixed by a separate PR: 

https://github.com/rust-lang/rust/pull/27087

**Pull request closed without merging**